### PR TITLE
Split functionality into two, distinct classes

### PR DESCRIPTION
**Pull request by ekampp**
One (Decoder) for decoding and one (Encoder) for encoding.

We should consider moving all the classes into a namespace, to we don't have namespace collisions with any other, potential gems.

### CODE REVIEW

**Comment by kholbekj** [comment]
I'm not sure how much i like the calling syntax of this structure. 
So, first up we change
`BEncoder.encode` and
`BEncoder.decode`
to 
`Encoder.encode` and
`Decoder.parse`
and then realise the need for a namespace, for instance 'BEncoder'
`BEncoder::Encoder.encode`
`BEncoder::Decoder.parse`
Which, to me, seems very extensive, compared to the scope of project.
Though I do agree that some level of seperation would be good, I'm not sure I think they should be called from different classes.

**Comment by kholbekj** [comment]
Oh wait, I'm not even right, you've removed the class-level methods? So it's actually
`BEncoder::Encoder.new(object).encode` and so forth?

**Comment by ekampp** [comment]
I have most certainly have not removed the class level methods. This way, you can call it exactly as you did Before: `BEncode.decode` or `BEncode.encode` and namespacing it will not necessarily lead to any change in this behavior.

File: lib/bencoder.rb [hunk 1]
```diff
@@ -3,136 +3,162 @@
...
+    when 'l'
+      parse_list string
+    when 'd'
+      parse_dict string
+    else
+      parse_string string
     end
+  end
 
+protected
```
Comment (kholbekj) [inline lib/bencoder.rb:30]: Why protected? These are as far as I can see private methods.

Comment (ekampp) [inline lib/bencoder.rb:30]: They shouldn't be shared with any potential child classes, imho, thus `protected`

Comment (ekampp) [inline lib/bencoder.rb:30]: But I don't have strong opinions about it.

**Pull request closed without merging**